Sources/GRPCCore/Call/Server/ServerInterceptorOperation.swift

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/// A `ServerInterceptorOperation` describes to which RPCs a server interceptor should be applied.
+///
+/// You can configure a server interceptor to be applied to:
+/// - all RPCs and services;
+/// - requests directed only to specific services registered with your server; or
+/// - requests directed only to specific methods (of a specific service).
+///
+/// - SeeAlso: ``ServerInterceptor`` for more information on server interceptors, and
+///  ``ClientInterceptorOperation`` for the client-side version of this type.
+public struct ServerInterceptorOperation: Sendable {
+  internal enum Wrapped: Sendable {
+    case allServices(interceptor: any ServerInterceptor)
+    case serviceSpecific(interceptor: any ServerInterceptor, services: [String])
+    case methodSpecific(interceptor: any ServerInterceptor, methods: [MethodDescriptor])
+  }
+
+  /// An operation specifying an interceptor that applies to all RPCs across all services will be registered with this server.
+  /// - Parameter interceptor: The interceptor to register with the server.
+  /// - Returns: A ``ServerInterceptorOperation``.
+  public static func applyToAllServices(
+    _ interceptor: any ServerInterceptor
+  ) -> Self {
+    Self(wrapped: .allServices(interceptor: interceptor))
+  }
+
+  /// An operation specifying an interceptor that will be applied only to RPCs directed to the specified services.
+  /// - Parameters:
+  ///   - interceptor: The interceptor to register with the server.
+  ///   - services: The list of service names for which this interceptor should intercept RPCs.
+  /// - Returns: A ``ServerInterceptorOperation``.
+  public static func apply(
+    _ interceptor: any ServerInterceptor,
+    onlyToServices services: [String]
```

Comment:
Also: we have a `ServiceDescriptor` – we should use it instead of `String` here

Reply:
Should the `service` in `MethodDescriptor` be a `ServiceDescriptor` then, instead of a `String`?
This is the main reason why I chose a `String` for this API.

Reply:
Probably, yes, but I think that'd have to come as part of a wider change though.

There are a couple of things that'd have to be considered before doing that:
- generated code would need updating
- parsing the ":path" would need to decompose the fully qualified service name (i.e. package + service) or have `ServiceDescriptor` do that lazily (store the fully qualified string, and compute the package/service name separately on demand)

Either way, I think `ServiceDescriptor` is the right type here and we should make any change to `MethodDescriptor` separately.

Reply:
Yes, I wasn't thinking we should do this as part of this PR, but we should probably make this change too for consistency.